### PR TITLE
RNN : Support for nD data

### PR DESCRIPTION
**Pull request by farizrahman4u**
Currently,  only 3D input is supported by the rnn function.

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -427,7 +427,8 @@ def rnn(step_function, inputs, output_dim, initial_states,
         new_states: list of tensors, latest states returned by
             the step function, of shape (samples, ...).
     '''
-    inputs = tf.transpose(inputs, (1, 0, 2))
+    axes = [1, 0] + range(2, len(inputs.get_shape()))
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: I think it would be good to assert than the input tensor is at least 3D (to avoid a more obscure error message later).

**Comment by farizrahman4u** [comment]
Passing 2D data  wouldn't break anything (If the step function expects 1D at each eteps). And there are assertions already at the `Recurrent` layer level.

**Comment by fchollet** [comment]
> And there are assertions already at the Recurrent layer level.

Yes, but the point of the `K.rnn` API is to be used anywhere (e.g. it's used in `layers.core`, `activations`...), including custom user code.

**Comment by farizrahman4u** [comment]
Agreed. But what if the input shape is `(samples, timesteps)` (like a continuous mask over timesteps for another rnn). If we assert `ndim>=3` then the user will be forced to reshape to `(samples, timesteps, 1)`  right? So we should be asserting `ndim>=2` isn't it?

**Comment by farizrahman4u** [comment]
Agreed, done.

**Comment by fchollet** [comment]
Cool, thanks. Btw it's "at least" in 2 words. Otherwise the PR looks good to me. Please squash your commits and I will merge.

**Comment by farizrahman4u** [comment]
Squashed.

**Comment by farizrahman4u** [comment]
@fchollet Congratz!! 4000 stars. :+1:

**Comment by fchollet** [comment]
> @fchollet Congratz!! 4000 stars. :+1:

At this rate, we will get to `n` stars in finite time.

**Comment by fchollet** [comment]
Tests are passing locally. Everything looks good.
